src/contract/CallParams.ts

```diff
@@ -0,0 +1,751 @@
+
+        return this.addParam(param);
+    }
+
+    public addUint256(param: Uint8Array): this {
+        this.conditionallyAddType({ ty: ArgumentType.uint256, array: false });
+
+        return this.addParam(param);
+    }
+
+    public addUint256Array(param: Uint8Array[]): this {
+        this.conditionallyAddType({ ty: ArgumentType.uint256, array: true });
+
+        return this.addParam(param);
+    }
+
+    public addString(param: string): this {
+        this.conditionallyAddType({ ty: ArgumentType.string, array: false });
+
+        return this.addParam(param);
+    }
+
+    public addAddress(param: Uint8Array | string): this {
+        const par = param instanceof Uint8Array ?
+            param :
+            Buffer.from(param as string, "hex");
+
+        if (par.length !== 20) {
+            throw new Error("`address` type requires parameter to be exactly 20 bytes");
+        }
+
+        this.conditionallyAddType({ ty: ArgumentType.address, array: false });
+
+        return this.addParam(par);
+    }
+
+    public addBytes(param: Uint8Array): this {
+        this.conditionallyAddType({ ty: ArgumentType.bytes, array: false });
+
+        return this.addParam(param);
+    }
+
+    public addBytesFixed(param: Uint8Array): this {
+        if (param.length > 32) {
+            throw new Error("`bytes<M>` type requires parameter length to be less than or equal to 32 bytes");
+        }
+
+        this.conditionallyAddType({ ty: ArgumentType.bytesfix, array: false }, param.length);
+
+        return this.addParam(param);
+    }
+
+    public addFunction(address: Uint8Array | string, func: FunctionSelector | string): this {
+        const addressParam = address instanceof Uint8Array ?
+            address :
+            Buffer.from(address as string, "hex");
+
+        const functionSelector = func instanceof FunctionSelector ?
+            (func as FunctionSelector).toProto() :
+            FunctionSelector.identifier(func as string);
+
+        if (addressParam.length !== 20) {
+            throw new Error("`function` type requires parameter `address` to be exactly 20 bytes");
+        }
+
+        this.conditionallyAddType({ ty: ArgumentType.func, array: false });
+
+        const proto = new Uint8Array(24);
+        proto.set(addressParam, 0);
+        proto.set(functionSelector, 20);
+
+        return this.addParam(proto);
+    }
+
+    public addStringArray(strArray: string[]): this {
+        this.conditionallyAddType({ ty: ArgumentType.string, array: true });
+
+        return this.addParam(strArray);
+    }
+
+    public addByteArray(byteArray: Uint8Array[]): this {
+        this.conditionallyAddType({ ty: ArgumentType.bytes, array: true });
+
+        return this.addParam(byteArray);
+    }
+
+    public addAddressArray(addresses: string[] | Uint8Array[]): this {
+        if (!Array.isArray(addresses)) {
+            throw new TypeError("string[] and Uint8Array[] are both arrays, how is this possible?");
```

Comment:
This is possible if this API is used improperly from a pure Javascript context (i.e. without type-checking). I think we need to have a discussion on whether we want to have these kinds of assertions everywhere or to simply rely on type annotations, because it's pretty inconsistent throughout the SDK.

I would say we should only have type assertions if it's necessary to satisfy `tsc`, otherwise it's on the user to provide the correct types (as is the case in most, but not all, JS APIs).

At the very least, this error should be more authoritative (`expected array, got ${typeof addresses}` or something).